[mlir][bufferization] Generalize tensor slice rules to subset ops

matthias-springer opened the pull request:
This commit generalizes the special tensor.extract_slice/tensor.insert_slice bufferization rules to tensor subset ops.

Ops that insert a tensor into a tensor at a specified subset (e.g., tensor.insert_slice, tensor.scatter) can implement the `SubsetInsertionOpInterface`.

Apart from adding a new op interface (extending the API), this change is NFC. The only ops that currently implement the new interface are tensor.insert_slice and tensor.parallel_insert_slice, and those ops were are supported by One-Shot Bufferize.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td:
```diff
@@ -0,0 +1,118 @@
+    op. E.g., "tensor.insert_slice" defines the subset via a hyperrectangular
+    slice. A scatter operation could define the subset via a list of indices.
+
+    Ops that deal with tensor subsets come in two flavours:
+    - Insertion flavor: Ops that insert a source tensor into a destination
+      tensor at the specified subset. Such ops usually return a new destination
+      tensor and implement the `DestinationStyleOpInterface`. Insertion ops can
+      implement the `SubsetOpInterface`. Example: "tensor.insert_slice"
+    - Extraction flavor: Ops that define a tensor subset. They extract a
+      specified subset from a tensor. There is currently no op interface for
+      such ops. Example: "tensor.extract_slice"
```
Renamed to `SubsetInsertionOpInterface`.

matthias-springer commented on mlir/lib/Dialect/Tensor/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -1172,6 +1107,11 @@ void mlir::tensor::registerBufferizableOpInterfaceExternalModels(
     ReshapeOp::attachInterface<ReshapeOpInterface>(*ctx);
     SplatOp::attachInterface<SplatOpInterface>(*ctx);
 
+    // SubsetOpInterface models.
+    InsertSliceOp::attachInterface<InsertSliceOpSubsetOpInterface>(*ctx);
+    ParallelInsertSliceOp::attachInterface<
+        ParallelInsertSliceOpSubsetOpInterface>(*ctx);
+
```
It is only used in bufferization at the moment and I wanted to make sure that this change is NFC for users for the moment. I.e., they don't have to add an additional `registerSubsetOpInterfaceExternalModels` call, but registering the `BufferizableOpInterface` models is enough. If we use this interface for other transforms (tiling, etc.) in the future, I'm going to move it to a separate file.

maerhart replied:
I see, what about moving the interface implementation into a separate `SubsetInsertionOpInterfaceImpl.cpp` already, exposing a dedicated registration function in the public API, but then calling that in the bufferizableOpInteface registration function?

It just feels a bit inconsistent to me to put the concrete implementation in the same file but have a separate td and header file for the interface declaration.

matthias-springer replied:
Moved to separate file and registering the models together with BufferizableOpInterface models.

maerhart replied:
That looks great, thanks!

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td:
````diff
@@ -0,0 +1,118 @@
+              $_op.getOperation());
+        }]
+      >,
+      InterfaceMethod<
+        /*desc=*/[{
+          Return "true" if this operation inserts into a subset that is
+          equivalent to the subset defined by `candidate`.
+
+          Two subsets are "equivalent" and "same" if they can bufferize to the
+          same buffer views/aliases. If they are "equivalent", the tensor IR
+          may be expressed in terms of different SSA values (but they could
+          bufferize to MemRef SSA values that can CSE without breaking
+          correctness). `equivalenceFn` should return "true" if the two given
+          values are equivalent.
+
+          Example:
+          ```
+          // The subset of the SubsetOpInterface op %1 is equivalent to the
+          // subset defined by %2 (but not "same"):
+          %0 = arith.select %c, %t, %t : tensor<?xf32>
+          %1 = tensor.insert_slice %x into %0[0][5][1]
+              : tensor<5xf32> into tensor<?xf32>
+          %2 = tensor.extract_slice %t[0][5][1] : tensor<?xf32> to tensor<5xf32>
+
+          // The subset of the SubsetOpInterface op %1 is equivalent to and
+          // "same" as the subset defined by %2.
+          %1 = tensor.insert_slice %x into %t[0][5][1]
+              : tensor<5xf32> into tensor<?xf32>
+          %2 = tensor.extract_slice %t[0][5][1] : tensor<?xf32> to tensor<5xf32>
+          ```
+        }],
+        /*retType=*/"bool",
+        /*methodName=*/"isEquivalentSubset",
+        /*args=*/(ins
+            "::mlir::Value":$candidate,
+            "::llvm::function_ref<bool(Value, Value)>":$equivalenceFn)
+      >,
````
The implementation also has to compare offsets, sizes and strides. For other ops such as `tensor.scatter`/`gather`, it will have to compare the tensor of indices. What must be compared depends on the op, that's why it's an interface method.

maerhart replied:
That makes sense 👍

maerhart commented on mlir/lib/Dialect/Tensor/Transforms/SubsetInsertionOpInterfaceImpl.cpp:
```diff
@@ -0,0 +1,85 @@
+using namespace mlir::bufferization;
+using namespace mlir::tensor;
+
+namespace mlir {
+namespace tensor {
+namespace {
```
Why nest an anonymous namespace in mlir/tensor when there is already a `using namespace` for it?

matthias-springer replied:
To make sure that the function is not visible outside of this file. It's the same as making the functions `static`.

matthias-springer committed:
[mlir][bufferization] Generalize tensor slice rules to subset ops

This commit generalizes the special tensor.extract_slice/tensor.insert_slice bufferization rules to tensor subset ops.

Ops that insert a tensor into a tensor at a specified subset (e.g., tensor.insert_slice, tensor.scatter) can implement the `SubsetOpInterface`.

Apart from adding a new op interface (extending the API), this change is NFC. The only ops that currently implement the new interface are tensor.insert_slice and tensor.parallel_insert_slice, and those ops were are supported by One-Shot Bufferize.
